lib/internal/mime.js

```diff
@@ -0,0 +1,318 @@
+'use strict';
+
+const {
+  ArrayPrototypeJoin,
+  ObjectCreate,
+  MapIteratorPrototypeNext,
+  RegExpPrototypeTest,
+  RegExpPrototypeExec,
+  StringPrototypeReplace,
+  StringPrototypeToLowerCase,
+  StringPrototypeSlice,
+  StringPrototypeCharAt,
+  SymbolReplace,
+  SymbolIterator,
+  SafeMap: Map,
+} = primordials;
+
+const NotHTTPTokenCodePoint = /[^!#$%&'*+\-.^_`|~A-Za-z0-9]/g;
+const NotHTTPQuotedStringCodePoint = /[^\t\u0020-~\u0080-\u00FF]/g;
+NotHTTPQuotedStringCodePoint[SymbolReplace] = null;
+
+const HTTP_TAB_OR_SPACE = '\t ';
+const HTTP_WHITESPACE = ArrayPrototypeJoin(['\r\n', ...HTTP_TAB_OR_SPACE], '');
+const BEGINNING_WHITESPACE = new RegExp(
+  `^[${StringPrototypeReplace(HTTP_WHITESPACE, /\W/g, '\\$&')}]*`);
+const ENDING_WHITESPACE = new RegExp(
+  `[${StringPrototypeReplace(HTTP_WHITESPACE, /\W/g, '\\$&')}]*$`);
+
+function toASCIILower(str) {
+  return StringPrototypeReplace(str, /[A-Z]/g, (c) => StringPrototypeToLowerCase(c));
+}
+
+function parseTypeAndSubtype(str) {
+  // Skip only HTTP whitespace from start
+  str = StringPrototypeReplace(str, BEGINNING_WHITESPACE, '');
+  let position = 0;
+  // read until '/'
+  const typeMatch = execPattern(/^[^/]*/, str);
+  if (!typeMatch ||
+    typeMatch[0] === '' ||
+    testPattern(NotHTTPTokenCodePoint, typeMatch[0])) {
+    throw new TypeError('Invalid MIME type');
+  }
+  position = typeMatch[0].length;
+  // skip '/'
+  position += 1;
+  const type = toASCIILower(typeMatch[0]);
+  // read until ';'
+  const subtypeMatch = execPattern(
+    /^[^;]*/,
+    StringPrototypeSlice(str, position));
+  if (subtypeMatch) {
+    position += subtypeMatch[0].length;
+    // skip ';'
+    position += 1;
+    subtypeMatch[0] = StringPrototypeReplace(
+      subtypeMatch[0],
+      ENDING_WHITESPACE,
+      '');
+  }
+  if (!subtypeMatch ||
+    subtypeMatch[0] === '' ||
+    testPattern(NotHTTPTokenCodePoint, subtypeMatch[0])) {
+    throw new TypeError('Invalid MIME subtype');
+  }
+  const subtype = toASCIILower(subtypeMatch[0]);
+  const target = ObjectCreate(null);
+  target.type = type;
+  target.subtype = subtype;
+  target.parametersString = StringPrototypeSlice(str, position);
+  return target;
+}
+
+function parseParametersString(str) {
+  const paramsMap = new Map();
+  str = StringPrototypeReplace(str, ENDING_WHITESPACE, '');
+  let position = 0;
+  while (position < str.length) {
+    // Skip any whitespace before parameter
+    const skip = execPattern(
+      BEGINNING_WHITESPACE,
+      StringPrototypeSlice(str, position))[0];
+    if (skip) position += skip.length;
+    // Read until ';' or '='
+    const parameterMatch = execPattern(
+      /^[^;=]*/,
```

Comment:
Nit: can be replaced with a simple string search (this won't work with unicode symbols of surrogate pairs but i don't think it matters given we will only use ascii search characters IIUC)
i.e.
```js
const indexOfAny = (str, chars, offset = 0) => {
// or maybe have `chars` as Set and use .has()
  let index = ofset;  
  while (index < str.length && !chars.includes(str[index]))
    index++;
  return index === str.length ? -1 : index;
}
// and then here it will be
const parameterEnd = indexOfAny(StringPrototypeSlice(str, position), ';=');
position += parameterEnd !== -1 ? parameterEnd : 0;
```

Reply:
we do not need unicode code point searching, correct.

this reads as more complicated and is much bigger code wise than using the regular expression.